macaroon_service_test.go

```diff
@@ -94,6 +96,122 @@ func TestMacaroonServiceMigration(t *testing.T) {
 	require.NoError(t, ms.Stop())
 }
 
+// TestMacaroonGeneration tests that the macaroon service generates macaroons
+// properly, and doesn't regenerate the macaroon if the required permissions
+// haven't changed, but that it also does regenerate the macaroon when required
+// as the required permissions change.
+func TestMacaroonGeneration(t *testing.T) {
+	// Create a temporary directory where we can store the macaroon db and
+	// the macaroon we are about to create.
+	tempDirPath, err := ioutil.TempDir("", ".testMacaroons")
+	require.NoError(t, err)
+	defer os.RemoveAll(tempDirPath)
+
+	// Set the macaroon store, and get the root key store.
+	rks, backend, err := NewBoltMacaroonStore(
+		tempDirPath, "macaroons.db", defaultDBTimeout,
+	)
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, backend.Close())
+	}()
+
+	// Create the initial config. We set a mock DB password, as we either
+	// need to set a DB password or provide an LndClient, ephemeral key and
+	// key locator so that a shared key can be derived with LND, when
+	// creating the macaroon service.
+	cfg := &MacaroonServiceConfig{
+		MacaroonLocation: tempDirPath,
+		DBPassword:       []byte{1, 1, 1, 1, 1, 1, 1, 1, 1, 1},
+		RootKeyStore:     rks,
+		MacaroonPath:     filepath.Join(tempDirPath, "test.macaroon"),
+	}
+
+	// Generate mock permissions.
+	firstEntityWrite := bakery.Op{Entity: "first_entity", Action: "write"}
+	firstEntityRead := bakery.Op{Entity: "first_entity", Action: "read"}
+	secondEntityRead := bakery.Op{Entity: "second_entity", Action: "read"}
+
+	mockPerms := make(map[string][]bakery.Op)
+
+	// 1. We start by adding 2 permissions to the mockPerms map, and ensure
+	// that the macaroon service generates a macaroon with the correct
+	// permissions.
+	mockPerms["/uri1.Test/WriteTest"] = []bakery.Op{firstEntityWrite}
+	mockPerms["/uri1.Test/ReadTest"] = []bakery.Op{firstEntityRead}
+	cfg.RequiredPerms = mockPerms
+
+	ms := generateMacaroonService(t, cfg)
+
+	// Ensure that the macaroon service generates a macaroon with the
+	// correct permissions.
+	mac1, mac1Bytes := extractMacaroon(t, cfg.MacaroonPath)
+	ensureMacaroonPermissions(t, mac1, mockPerms, ms)
+
+	require.NoError(t, ms.Stop())
+
+	// 2. Secondly, we create a new macaroon service with the same
+	// permissions as the first macaroon service, and ensure that the
+	// macaroon hasn't been regenerated.
+	ms = generateMacaroonService(t, cfg)
```

Comment:
you can just keep stopping and starting the same macaroon service. No need to keep generating a new one.